src/Admin/AbstractAdmin.php

```diff
@@ -2849,6 +2855,11 @@ public function canAccessObject($action, $object)
         return $object && $this->id($object) && $this->hasAccess($action, $object);
     }
 
+    protected function configureQuery(ProxyQueryInterface $query): ProxyQueryInterface
+    {
+        return $query;
+    }
```

Comment:
What if a contributor adds more code here, and we forget they shouldn't? Same issue. Maybe we should just add an interface and check if it is implemented? I'm on my phone, but if you have a look at the issues you will find some where I write about the big pile of SRP this class is, and some plans to avoid adding to it. Since you are contributing a lot (thanks, BTW), it's better if you have this issue in mind.

Reply:
https://github.com/sonata-project/SonataAdminBundle/issues/3947

Reply:
This is why there is protected function VS public function. And not final functions VS final functions.

For example there is the following function with no code inside.
```
protected function configureFormFields(FormMapper $form)
{
}
```
called in the public `defineFormBuilder`

I expect from Sonata to not write code in the configureFormFields but in the function `defineFormBuilder`.

What about having two AbstractAdmin @greg0ire ?

One AbstractBaseAdmin (or any better name) implementing the interface and listing abstract protected function needed for configuration.
There should be no reason for the user to override theses methods.

One AbstractAdmin extending the AbstractBaseAdmin, and implementing only the protected function with empty/basic function.
There should be no reason for the sonata-contributors to write code in theses methods.

This is BC.
AbstractBaseAdmin

And in the next major, all the method from the AbstractBaseAdmin will be final.
If overriding a method in the AbstractBaseAdmin is needed, create a protected and implement it in the AbstractBaseAdmin.

It could be an easy and BC way to start.
And then, maybe some others improvements will come to mind.

Reply:
> One AbstractBaseAdmin (or any better name) implementing the interface and listing abstract protected function needed for configuration.
There should be no reason for the user to override theses methods.

If it's a list of abstract function then it's not really an implementation, more like a big interface, just like `AdminInterface` is (although I split it in #4440). To me, there should not be the `extends AccessRegistryInterface, FieldDescriptionRegistryInterface, LifecycleHookProviderInterface, MenuBuilderInterface, ParentAdminInterface, UrlGeneratorInterface` part of `AdminInterface` in the future (and ideally, that interface and the `AbstractAdmin` should both disappear).

Reply:
> If it's a list of abstract function then it's not really an implementation, more like a big interface

There would be more than just abstract function.

If I take a look at the actual AbstractAdmin functions.
- getExportFormats => There should be a way to override the exportFormats, so it should stay. (But it seems weird that the comment say `NEXT_MAJOR: return null to indicate no override`)
- getExportFields => There should be a way to override the exportFields, it should stay (but there is something about `configureExportFields` maybe it's better, but I don't know how it works).
- getDataSourceIterator => I don't think it's usefull to override this, this should go in AbstractBaseAdmin.
- validate => Stay
- initialize => Go in BaseAdmin since it call an empty function configure.
- configure => Stay
- update => Go in BaseAdmin since it call preUpdate/postUpdate
- preUpdate/postUpdate => Stay

Etc.

Reply:
initialize call configure
update call preUpdate and postUpdate
create call PrePersist and postPersist
delete call PreRemove and postRemove
buildForm call configureFormFields
buildList call configureListFields
buildShow call configureShowFields

initialize, update, create, delete, buildForm, buildList, buildShow should be final then.

etc.
In the same way, I don't think as bad the fact that createQuery call a configureQuery function and that createQuery will be final.

I need to fix the tests, but I'm not sure we have to find the answer to your question "How to split the AbstractAdmin" in order to decide if yes or no it's a good PR. :)

Reply:
> I need to fix the tests, but I'm not sure we have to find the answer to your question "How to split the AbstractAdmin" in order to decide if yes or no it's a good PR. :)

Yeah no, definitely not, it was more of a "big picture" discussion, consider it independent from this PR :P

Reply:
BTW, I see a lot of function in common between AbstractAdminExtension and AbstractAdmin. I don't know what is an AdminExtension, but maybe there is something to do with this ?
Like removing from the AbstractAdmin all the function defined in AbstractAdminExtension and asking the developper to create an AdminExtension for each admin ?

Reply:
And admin extension is a way to extend an admin by using composition, without resorting to inheritance at all. Well… if you don't extend from `AbstractAdminExtension`, that is :rofl: 
The feature is documented here: https://sonata-project.org/bundles/admin/3-x/doc/reference/extensions.html and relies on implementing an interface that clearly violates the ISP (if it weren't you wouldn't need `AbstractAdminExtension`). IMO this is not the way, IMO the way is to have small objects with a defined purpose that make up the admin (like the `BreadcrumbsBuilder` that I extracted, for instance: you might not know that class, but I bet you have an idea about its purpose just by reading the name). The `TemplateRegistry` is another good example.

Reply:
> And admin extension is a way to extend an admin by using composition, without resorting to inheritance at all. Well… if you don't extend from `AbstractAdminExtension`, that is 🤣

I always used inheritance 😅